File: src/main/java/org/radarbase/appserver/service/GithubClient.java
/*
 *
 *  *
 *  *  * Copyright 2018 King's College London
 *  *  *
 *  *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  *  * you may not use this file except in compliance with the License.
 *  *  * You may obtain a copy of the License at
 *  *  *
 *  *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *  *
 *  *  * Unless required by applicable law or agreed to in writing, software
 *  *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  * See the License for the specific language governing permissions and
 *  *  * limitations under the License.
 *  *  *
 *  *
 *
 */

package org.radarbase.appserver.service;

import jakarta.annotation.Nonnull;
import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.beans.factory.config.ConfigurableBeanFactory;
import org.springframework.context.annotation.Scope;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Component;
import org.springframework.web.server.ResponseStatusException;
import org.springframework.web.util.UriComponentsBuilder;

import java.io.IOException;
import java.io.InputStream;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.concurrent.Executor;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

@Slf4j
@Component
@Scope(value = ConfigurableBeanFactory.SCOPE_SINGLETON)
public class GithubClient {
    private static final String GITHUB_API_URI = "api.github.com";
    private static final String GITHUB_API_ACCEPT_HEADER = "application/vnd.github.v3+json";

    @Nonnull
    private final transient String authorizationHeader;

    private transient final Duration httpTimeout;
    private transient final HttpClient client;

    @Value("${security.github.client.maxContentLength:1000000}")
    private transient int maxContentLength;

    @SneakyThrows
    @Autowired
    public GithubClient(
            @Value("${security.github.client.timeout:10}") int httpTimeout,
            @Value("${security.github.client.token:}") String githubToken) {
        this.authorizationHeader = githubToken != null ? "Bearer " + githubToken.trim() : "";
        this.httpTimeout = Duration.ofSeconds(httpTimeout);
        this.client = HttpClient.newBuilder()
                .version(HttpClient.Version.HTTP_1_1)
                .followRedirects(HttpClient.Redirect.NORMAL)
                .connectTimeout(this.httpTimeout)
                .build();
    }

    public String getGithubContent(String url) throws IOException, InterruptedException {
        log.debug("Fetching Github URL {}", url);
        URI uri = URI.create(url);
        HttpResponse<InputStream> response = makeRequest(uri);

        if (response.statusCode() >= 200 && response.statusCode() < 300) {
            checkContentLengthHeader(response);

            try (InputStream inputStream = response.body()) {
                byte[] bytes = inputStream.readNBytes(maxContentLength + 1);
                checkContentLength(bytes.length);
                return new String(bytes);
            }
        } else {
            log.error("Error getting Github content from URL {} : {}", url, response);
            throw new ResponseStatusException(
                    HttpStatus.valueOf(response.statusCode()), "Github content could not be retrieved");
        }
    }

    private HttpResponse<InputStream> makeRequest(URI uri) throws InterruptedException {
        try {
            return client.send(getRequest(uri), HttpResponse.BodyHandlers.ofInputStream());
        } catch (IOException ex) {
            log.error("Failed to retrieve data from github {}: {}", uri, ex.toString());
            throw new ResponseStatusException(HttpStatus.BAD_GATEWAY, "Github responded with an error.");
        }
    }

    private void checkContentLengthHeader(HttpResponse<?> response) {
        response.headers().firstValue("Content-Length")
                .map((l) -> {
                    try {
                        return Integer.valueOf(l);
                    } catch (NumberFormatException ex) {
                        return null;
                    }
                })
                .ifPresent(this::checkContentLength);
    }

    private void checkContentLength(int contentLength) {
        if (contentLength > maxContentLength) {
            throw new ResponseStatusException(
                    HttpStatus.BAD_REQUEST, "Github content is too large");
        }
    }

    public URI getValidGithubUri(URI uri) throws IOException {
        if (uri.getHost().equalsIgnoreCase(GITHUB_API_URI)
                && uri.getScheme().equalsIgnoreCase("https")
                && (uri.getPort() == -1 || uri.getPort() == 443)) {
                    return UriComponentsBuilder.newInstance()
                        .scheme("https").host(uri.getHost()).path(uri.getPath()).query(uri.getQuery()).build().toUri();
                }
        else {
            throw new MalformedURLException("Invalid Github url.");
        }
    }

    private HttpRequest getRequest(URI uri) throws IOException {
        URI validUri = this.getValidGithubUri(uri);
        HttpRequest.Builder request = HttpRequest.newBuilder(validUri)
                .header("Accept", GITHUB_API_ACCEPT_HEADER)
                .GET()
                .timeout(httpTimeout);
        if (!authorizationHeader.isEmpty()) {
            request.header("Authorization", authorizationHeader);
        }
        return request.build();
    }
}

File: src/main/java/org/radarbase/appserver/service/GithubClient.kt
/*
 *
 *  *
 *  *  * Copyright 2018 King's College London
 *  *  *
 *  *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  *  * you may not use this file except in compliance with the License.
 *  *  * You may obtain a copy of the License at
 *  *  *
 *  *  *   http://www.apache.org/licenses/LICENSE-2.0
 *  *  *
 *  *  * Unless required by applicable law or agreed to in writing, software
 *  *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  *  * See the License for the specific language governing permissions and
 *  *  * limitations under the License.
 *  *  *
 *  *
 *
 */

package org.radarbase.appserver.service

import io.ktor.client.*
import io.ktor.client.call.*
import io.ktor.client.engine.cio.*
import io.ktor.client.plugins.defaultRequest
import io.ktor.client.request.*
import io.ktor.http.*
import kotlinx.coroutines.runBlocking
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Value
import org.springframework.beans.factory.config.ConfigurableBeanFactory
import org.springframework.context.annotation.Scope
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Component
import org.springframework.web.server.ResponseStatusException
import org.springframework.web.util.UriComponentsBuilder
import java.io.IOException
import java.net.MalformedURLException
import java.net.URI

/**
 * A singleton component responsible for interacting with the GitHub API.
 * Provides utility methods for authorized and unauthorized access to GitHub content.
 * The configuration for this client can be customized through application properties.
 *
 * @property httpTimeout The timeout in seconds for HTTP requests.
 * @property githubToken The personal access token for GitHub API authentication.
 */
@Component
@Scope(value = ConfigurableBeanFactory.SCOPE_SINGLETON)
class GithubClient(
    @Value("\${security.github.client.maxContentLength:1000000}")
    @Transient
    private val maxContentLength: Int,

    @Value("\${security.github.client.timeout:10}") httpTimeout: Int,
    @Value("\${security.github.client.token:}") githubToken: String
) {

    @Transient
    private val authorizationHeader: String =
        if (githubToken.isEmpty()) "" else "Bearer " + githubToken.trim { it <= ' ' }

    /**
     * An HTTP client configured to interact with the GitHub API using the CIO engine.
     * This client sets default request headers, specifies request timeout, and
     * enables automatic redirection following.
     */
    @Transient
    private val client = HttpClient(CIO) {
        defaultRequest {
            header(HttpHeaders.Accept, GITHUB_API_ACCEPT_HEADER)
        }
        engine {
            requestTimeout = httpTimeout * 1000L
        }
        followRedirects = true
    }

    /**
     * Retrieves the content from a specified GitHub URL. This method attempts to make an authenticated request
     * if authentication is enabled. If unauthorized access occurs, it retries without authentication.
     *
     * @param url The URL of the GitHub resource to fetch content from.
     * @param authenticated Indicates whether the request should include an authentication header (default is true).
     * @return The content retrieved from the specified GitHub URL as a String.
     * @throws ResponseStatusException If the response indicates an error or the content size is too large.
     */
    fun getGithubContent(url: String, authenticated: Boolean = true): String = runBlocking {
        val request = client.prepareGet(getValidGithubUri(url)) {
            if (authenticated && authorizationHeader.isNotEmpty()) {
                header(HttpHeaders.Authorization, authorizationHeader)
            }
        }
        val response = request.execute()
        if (response.status.value in 200..299) {
            val contentLengthHeader = response.headers[HttpHeaders.ContentLength]?.toIntOrNull()
            contentLengthHeader?.let { checkContentLength(it) }

            val contentStream: ByteArray = response.body<ByteArray>()
            checkContentLength(contentStream.size)
            return@runBlocking String(contentStream)
        } else if (response.status.value == 401 && authenticated) {
            logger.warn("Unauthorized access to Github content from URL {}, retrying..", url)
            return@runBlocking getGithubContent(url, false)
        } else {
            logger.error("Error getting Github content from URL {} : {}", url, response)
            throw ResponseStatusException(
                HttpStatus.valueOf(response.status.value), "Github content could not be retrieved"
            )
        }
    }

    fun getGithubContent(url: String): String = getGithubContent(url, true)

    /**
     * Validates the given content length against a predefined maximum content length.
     * Throws a `ResponseStatusException` if the content length exceeds the allowed maximum.
     *
     * @param contentLength the length of the content to be validated
     */
    private fun checkContentLength(contentLength: Int) {
        if (contentLength > maxContentLength) {
            throw ResponseStatusException(
                HttpStatus.BAD_REQUEST, "Github content is too large"
            )
        }
    }

    /**
     * Validates and converts a given GitHub API URI into a standardized format.
     *
     * @param uri The GitHub API URI to validate and format.
     * @return The validated and properly formatted GitHub API URI.
     * @throws IOException If the provided URI is invalid or does not meet the required GitHub API standards.
     */
    @Throws(IOException::class)
    fun getValidGithubUri(uri: String): String {
        val parsedUri = URI(uri)
        if (parsedUri.host.equals(GITHUB_API_URI, ignoreCase = true) &&
            parsedUri.scheme.equals("https", ignoreCase = true) &&
            (parsedUri.port == -1 || parsedUri.port == 443)
        ) {
            return UriComponentsBuilder.newInstance()
                .scheme("https").host(parsedUri.host).path(parsedUri.path).query(parsedUri.query).build().toString()
        } else {
            throw MalformedURLException("Invalid Github URL: $uri")
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(GithubClient::class.java)

        private const val GITHUB_API_URI = "api.github.com"
        private const val GITHUB_API_ACCEPT_HEADER = "application/vnd.github.v3+json"
    }
}
